File: Algos-DataStructure/Java/Stacks/SimpleTextEditor.java
public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Stack<StringBuilder> stack = new Stack<>();
        int q = scanner.nextInt();
        scanner.nextLine();
        StringBuilder s = new StringBuilder();

        for (int i = 0; i < q; i++) {
            String[] query = scanner.nextLine().split(" ");
            int type = Integer.parseInt(query[0]);
            switch (type) {
                case 1:
                    stack.push(new StringBuilder(s));
                    s.append(query[1]);
                    break;
                case 2:
                    stack.push(new StringBuilder(s));
                    int k = Integer.parseInt(query[1]);
                    s.delete(s.length() - k, s.length());
                    break;
                case 3:
                    int index = Integer.parseInt(query[1]) - 1;
                    System.out.println(s.charAt(index));
                    break;
                case 4:
                    if (!stack.empty()) {
                        s = stack.pop();
                    }
                    break;
                default:
                    break;
            }
        }
        scanner.close();

}

File: Algos-DataStructure/JavaScript/Stacks/SimpleTextEditor.js
function processData(input) {
     //Enter your code here
    // 1 is to append string to the end of current string
    // 2 delete the last k chracter of string
    // 3 print the k'th character of string
    // 4 undo the last operation of 1/2
    const lines = input.split('\n');
    const q = parseInt(lines[0]);
    const queries = [];

    for (let i = 1; i <= q; i++) {
        queries.push(lines[i].split(' '));
    }

    let text = '';
    const stack = [];

    for (const query of queries) {
        const [type, value] = query;

        if (type === '1') {
            stack.push(text);
            text += value;
        } else if (type === '2') {
            stack.push(text);
            text = text.slice(0, -parseInt(value));
        } else if (type === '3') {
            console.log(text.charAt(parseInt(value) - 1));
        } else if (type === '4') {
            text = stack.pop();
        }
    }
}